### PR TITLE
Pagination and basic search

### PR DESCRIPTION
**Pull request by carkod**
## Done

- Pagination and sorting
- More data to display multiple pages
- Basic search (description and priority)

## QA

- Check out this feature branch
- Run the site using the command `./run serve` or `dotrun`
- View the site locally in your web browser at: http://0.0.0.0:8001/security/cve
- Query pagination `http://0.0.0.0:8001/security/cve?limit={page-size}` e.g. http://0.0.0.0:8001/security/cve?limit=1, should only show one element.
- Query pagination offset `http://0.0.0.0:8001/security/cve?offset={offset}` e.g. http://0.0.0.0:8001/security/cve?limit=5&offset=5, should show the second page with only 3 elements.
- The previous two values will assume `offset=0` and `limit=20` by default if no value passed (first page from the beginning). Therefore, those 2 parameters are required if you want to navigate to the next page (or if you don't see any CVEs because of lack of data). 
- Sort by date `http://0.0.0.0:8001/security/cve?order_by=oldest`, will sort `http://0.0.0.0:8001/security/cve?order=newest` by default. Accepted values `oldest` and `newest`.
- Search contents of description `http://0.0.0.0:8001/security/cve?search_desc={description}` e.g. http://0.0.0.0:8001/security/cve?search_desc={description}
- Multiple searches example http://0.0.0.0:8001/security/cve?search_desc=gitlab&priority=low
- New data structure for main page has been reverted. Planning to remodel the data for packages and release_status.
- Package search not included.

### CODE REVIEW

File: templates/security/cve/index.html [hunk 1]
```diff
@@ -716,4 +716,4 @@ <h2>Recent CVEs affecting Ubuntu</h2>
 {% include "shared/contextual_footers/_contextual_footer.html" %}
 {% endwith %}
 
-{% endblock %}
+{% endblock %}
```
Comment (carkod) [inline templates/security/cve/index.html:719]: This is probably because I copy pasted to avoid breaking changes in the conflicts when I was rebasing.

**Comment by jpmartinspt** [comment]
@carkod I'm not sure this is properly rebased.

File: static/sass/styles.scss [hunk 2]
```diff
@@ -639,4 +639,4 @@ summary {
 
 .is-bordered--top {
   @extend %vf-pseudo-border--top;
-}
```
Comment (carkod) [inline static/sass/styles.scss:642]: Again this is copied from base cve branch to avoid conflicts

File: webapp/security/views.py [hunk 3]
```diff
@@ -268,8 +268,53 @@ def api_create_notice():
...
+    offset = page * page_size - page_size
+
+    if page < 1 or 1 < page > total_pages:
+        flask.abort(404)
+
+    sort = asc if order_by == "oldest" else desc
+
+    # Search functionality
+
+    if search_id:
```
Comment (jpmartinspt) [inline webapp/security/views.py:290]: what is `search_id`?

File: static/sass/styles.scss (hunk 2 above)
Comment (nottrobin) [reply static/sass/styles.scss:642]: ?This doesn't make sense.... regardless of rebasing, if there's a `}` in this branch that isn't in the `cve` branch then either CSS was broken in the `cve` branch or it's broken in this branch. I'll see if I can discover which it is.

Comment (nottrobin) [reply static/sass/styles.scss:642]: Oh my bad, it's simply a newline character.

File: templates/security/cve/index.html (hunk 1 above)
Comment (nottrobin) [reply templates/security/cve/index.html:21]: Again just a newline character. Not a concern.

**Comment by nottrobin** [comment]
(I fixed up your QA steps to change `snap dotrun` to `dotrun`)

When you say go to `http://0.0.0.0:8001/?page={pagesize}`, I assume you must mean `http://0.0.0.0:8001/security/cve?page={pagesize}`? I'll update your QA steps.

When I try out doing any queries on `http://0.0.0.0:8001/security/cve`, I just see a blank page:

image

This is presumably because you're passing data through to the index page but you haven't made any changes to the template to display it. You really should add some minimal work to the template to show that data is being passed through to make this easily QAable.

File: webapp/security/views.py (hunk 3 above)
Comment (nottrobin) [reply webapp/security/views.py:290]: @carkod Let's have a quick catch-up after the standup this morning.

File: webapp/security/views.py [hunk 4]
```diff
@@ -268,8 +268,53 @@ def api_create_notice():
 # ===
 def cve_index():
 
-    list_cve = db_session.query(CVE).limit(10)
-    return flask.render_template("security/cve/index.html", list_cve=list_cve)
+    page = flask.request.args.get("page", default=1, type=int)
```
Comment (nottrobin) [inline webapp/security/views.py:271]: Would you mind using `limit` and `offset` as per https://pages.apigee.com/rs/apigee/images/api-design-ebook-2012-03.pdf ?

File: webapp/security/views.py [hunk 5]
```diff
@@ -268,8 +268,53 @@ def api_create_notice():
...
+    order_by = flask.request.args.get("order", type=str)
+    search_id = flask.request.args.get("id", type=str)
```
Comment (nottrobin) [inline webapp/security/views.py:273]: Remove this, as discussed

File: webapp/security/views.py (hunk 4 above)
Comment (nottrobin) [reply webapp/security/views.py:272]: Delete this, as discussed

File: webapp/security/views.py [hunk 6]
```diff
@@ -268,8 +268,67 @@ def api_create_notice():
...
+
+    # Search functionality
+    if search_desc and package and priority:
+        matched_desc = cves_query.filter(
+            and_(
+                CVE.description.ilike(f"%{search_desc}%"),
+                func.lower(CVE.priority) == func.lower(priority),
+            )
+        ).all()
+        print(matched_desc)
```
Comment (jpmartinspt) [inline webapp/security/views.py:291]: Remove the `print`

File: webapp/security/views.py [hunk 7]
```diff
@@ -268,8 +268,64 @@ def api_create_notice():
...
+        cves_query = cves_query.filter(
+            CVE.description.ilike(f"%{search_desc}%"),
+        )
+
+    # Pagination
+    total_results = cves_query.count()
+    total_pages = ceil(total_results / limit)
+
+    if offset > total_pages:
+        flask.abort(404)
```
Comment (nottrobin) [inline webapp/security/views.py:303]: Seeing a 404 page in this case is very surprising. If the offset is too large, please just return an empty list.

File: webapp/security/views.py [hunk 8]
```diff
@@ -268,8 +268,69 @@ def api_create_notice():
...
+    # Pagination
+    total_results = cves_query.count()
+    total_pages = ceil(total_results / limit)
+    page_first_result = offset + 1
+    page_last_result = offset + len(list_cve)
+
+    if offset > total_results:
+        list_cve = []
+        total_pages = None
+        total_results = None
```
Comment (nottrobin) [inline webapp/security/views.py:315]: Isn't `total_results` still valid? total_pages too.

**Pull request merged**